src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel_UpdateUI.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    internal partial class DocumentOutlineViewModel
+    {
+        private readonly record struct UIData(ExpansionOption ExpansionOption, SnapshotPoint? CaretPoint);
+
+        /// <summary>
+        /// Queue to batch up work to do to highlight the currently selected symbol node, expand/collapse nodes,
+        /// then update the UI.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<UIData> _updateUIQueue;
+
+        public void EnqueueUIUpdateTask(ExpansionOption expansionOption, SnapshotPoint? caretPoint)
+        {
+            _updateUIQueue.AddWork(new UIData(expansionOption, caretPoint), cancelExistingWork: true);
+        }
+
+        private async ValueTask UpdateUIAsync(ImmutableSegmentedList<UIData> options, CancellationToken cancellationToken)
+        {
+            var (expansion, caretPoint) = options.Last();
+            var model = await _filterAndSortQueue.WaitUntilCurrentBatchCompletesAsync().ConfigureAwait(false);
+            if (model is null)
+            {
+                return;
+            }
+
+            var documentSymbolUIItems = DocumentOutlineHelper.GetDocumentSymbolUIItems(model.DocumentSymbolData);
```

Comment:
document threading invariants here.

Reply:
there are none

Reply:
that needs to be documented :)   this says it's getting UI items, so there would normally be a concern about that needing to be on the UI thread.  If we don't have that issue, then we should have the code say that.  Otherwise future maintainers may not know and may be concerned this is incorrect.